File: examples/openmp/f90/mandel.f90
      program mandel
      use ppm

      implicit none

      integer, parameter               :: dp=selected_real_kind(13)
      complex(kind=dp) :: c, z
      integer          :: ix, iy, cnt
      integer          :: numpe, maxiter
      integer, parameter               :: width=800, height=800
      integer, dimension(width,height) :: iterations
      real(kind=dp)                    :: xmin, xmax, ymin, ymax, x, y, dx, dy
      external mytimer
      integer mandelprofiler(2), parprof(2), loopprof(2)
!$    integer OMP_GET_NUM_THREADS
        save mandelprofiler, parprof, loopprof
        call TAU_PROFILE_INIT()
        call TAU_PROFILE_TIMER(mandelprofiler, 'PROGRAM MANDEL')
        call TAU_PROFILE_TIMER(parprof, 'OpenMP Parallel Region')
        call TAU_PROFILE_TIMER(loopprof, 'OpenMP Loop')
        call TAU_PROFILE_START(mandelprofiler)
        call TAU_PROFILE_SET_NODE(0)


! --- init
!     -1.5 0.5 -1.0 1.0
!     -.59 -.54 -.58 -.53
!     -.65 -.4 .475 .725
!     -.65 -.5 .575 .725
!     -.59 -.56 .47 .5
      write(unit=*, fmt="(A)") "xmin xmax ymin ymax maxiter (Sample: -.59 -.56 .47 .5 216) ?"
      read (unit=*, fmt=*) xmin, xmax, ymin, ymax, maxiter

! --- initialization
      dx = (xmax - xmin) / width
      dy = (ymax - ymin) / height
      numpe = 1

! --- calculate mandelbrot set
      call mytimer(0)
!$OMP PARALLEL
      call TAU_PROFILE_START(parprof)
!$    numpe = OMP_GET_NUM_THREADS()
!$OMP DO PRIVATE(ix,x,y,c,z,cnt)
      do iy=1,height
        call TAU_PROFILE_START(loopprof)
        y = ymin + iy*dy
        do ix=1,width
          x = xmin + ix*dx
          c = cmplx(x,y)
          z = (0.0,0.0)
          cnt = 0
          do 
            z = z*z + c
            cnt = cnt+1
            if ( (dble(z)*dble(z) + dimag(z)*dimag(z)) >= 16 .or. &
                 cnt >= maxiter ) then
              exit
            end if
          end do
          iterations(ix,iy) = cnt
        end do
        call TAU_PROFILE_STOP(loopprof)
      end do
      call TAU_PROFILE_STOP(parprof)
!$OMP END PARALLEL
      call mytimer(numpe)

! --- generate ppm file
      write (unit=*, fmt="(A)") "generate ppm file..."
      call ppmwrite("mandel.ppm", iterations, maxiter)

      call TAU_PROFILE_STOP(mandelprofiler)
      end program mandel

File: examples/openmp/c/mandel.c
#include <stdio.h>
#include <stdlib.h>
#include <Profile/Profiler.h>

extern void mytimer_(int *);

#ifdef _OPENMP
#  include <omp.h>
#endif

#include "ppm.h"

field iterations;
int main(int argc, char *argv[]) {
  double xmin, xmax, ymin, ymax, dx, dy;
  int numpe, maxiter, ix, iy;
  TAU_PROFILE_TIMER(mt, "main()", "int (int, char **)", TAU_DEFAULT);
  TAU_PROFILE_SET_NODE(0);
  
  TAU_PROFILE_START(mt);

  if ( argc != 6 ) {
    fprintf(stderr, "%s: xmin xmax ymin ymax maxiter\n", argv[0]);
    fprintf(stderr, "Using defaults: -.59 -.56 .47 .5 216\n");
    xmin = -.59; xmax = -.56; ymin = .47; ymax = .5; maxiter = 216; 
  }
  else {
    xmin = strtod(argv[1], 0);
    xmax = strtod(argv[2], 0);
    ymin = strtod(argv[3], 0);
    ymax = strtod(argv[4], 0);
    maxiter =  atoi(argv[5]);
  }

  /* --- initialization -- */
  numpe = 1;
  dx = (xmax - xmin) / width;
  dy = (ymax - ymin) / height;

  /* --- calculate mandelbrot set -- */
  mytimer_(0);
#pragma omp parallel
  {
    TAU_PROFILE_TIMER(pt, "Parallel Region", " " , TAU_DEFAULT);
    TAU_PROFILE_START(pt);
#ifdef _OPENMP
    numpe = omp_get_num_threads();
#endif
#pragma omp for private(ix,iy)
  for (ix=0; ix<width; ++ix) {
    double x = xmin + ix*dx;
    TAU_PROFILE_TIMER(fl, "For loop", " " , TAU_DEFAULT);
    TAU_PROFILE_START(fl);
    for (iy=0; iy<height; ++iy) {
      double y = ymin + iy*dy;
      double zx, zy, cx, cy, ox, oy;
      int count;
      zx = 0.0; zy = 0.0;
      ox = 0.0; oy = 0.0;
      cx = x;   cy = y;
      count = 0;
      while ( (ox*ox + oy*oy) < 16 && count < maxiter ) {
	zx = ox*ox-oy*oy+cx;
        zy = ox*oy+ox*oy+cy;
        ++count;
	ox = zx;
	oy = zy;
      }
      iterations[ix][iy] = count;
    }
    TAU_PROFILE_STOP(fl);
  }
    TAU_PROFILE_STOP(pt);
  }
  mytimer_(&numpe);

  /* --- generate ppm file -- */
  printf("Writing picture ...\n");
  ppmwrite("mandel.ppm", iterations, maxiter);
  TAU_PROFILE_STOP(mt);
  exit(0);
}

